misk/src/main/kotlin/misk/backoff/Retries.kt

```diff
@@ -3,9 +3,17 @@ package misk.backoff
 /**
  * Retries the provided function up to a certain number of times, applying the given backoff
  * between each retry. The retry function is provided with current retry count, in case this is
- * relevant
+ * relevant.
+ *
+ * retryException returns true if the caught exception should be retried, subject to the limit.
+ * By default all exceptions are retried.
  */
-fun <A> retry(upTo: Int, withBackoff: Backoff, f: (retryCount: Int) -> A): A {
+fun <A> retry(
+  upTo: Int,
+  withBackoff: Backoff,
+  retryException: ((e: Exception) -> Boolean)? = null,
```

Comment:
rather than making the function optional, default it to a lambda that returns true?

Reply:
Actually I think this method is too complex and should be replaced with a class. A method that accepts multiple lambdas is a bad method.

I’m tempted to leave the basic retry thing here as-is, but to write a custom retry loop in Transacter because it has special requirements. It might be the only thing in our entire codebase that needs these bells and whistles, and it keeps the entire machine simpler if we do them directly.

Reply:
Ah, I used a default lambda originally. The code is cleaner that way, but I wondered about performance compared to a null-check.

This retry function is currently used for sending logs to CloudWatch, but doesn't have special requirements. I will rewrite the retry logic in Transacter. I think you're right it will be simpler to do this directly, particularly if there are more bells and whistles to add (such as "overriding" the backoff based on specific db failures). If more use cases show up that have similar requirements, we can consider DRYing it up.

Reply:
👍🏻